src/core/matcher.rkt

```diff
@@ -115,25 +116,27 @@
            (display ">" port))])
 
 (define (rewrite-expression expr #:destruct [destruct? #f] #:root [root-loc '()])
+  (define env (for/hash ([v (free-variables expr)]) (values v 'real)))
   (reap [sow]
-    (for ([rule (*rules*)])
+    (for ([rule (if (equal? 'complex (type-of expr env)) (*complex-rules*) (*rules*))])
```

Comment:
It's a bit of a shame that we end up recomputing the type of things over and over again, but for now I think that's fine. In the future, we can work on saving this information and not recomputing it.

Reply:
added a TODO here